### PR TITLE
Terrain format specification

### PR DESCRIPTION
**Pull request by heinezen**
Specification for the improved .terrain format discussed in #965 .

A few significant changes were made in comparison to version 0.

* **blending_priority**: Changed from its own attribute to a parameter of `blending_mask`. This allows us to define multiple blending mask with different priority values which could be useful for animated terrain textures.
* **dots_per_tile**: Now removed. I think we can leave this out if the engine is not tile-based anyway. Drawing the texture and blending should not require information about tiling.
* **scalefactor**: Included here with same purpose as in the sprite format (version 1).
* **blend_id** : Is now optional for a frame. If omitted, the blending mask of adjacent textures is always used.

### CODE REVIEW

File: doc/media/openage/terrain_format_spec.md [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+
+
+### `blendmask`
+
+Defines a blending mask that is used to blend one terrain texture
+into a different terrain texture that is directly adjacent. The
+blending mechanism is described in more detail in the [blendomatic](/doc/media/blendomatic.md)
+documentation. Only one of the blending masked defined by adjacent
+terrain is used (the one with the highest priority). Blending masks
+are **optional** and do not need to be defined.
```
Comment (TheJJ) [inline doc/media/openage/terrain_format_spec.md:132]: what's the content of a blendmask filename? it likely needs the same properties as tiles: texture offsets, and sizes

File: doc/media/openage/terrain_format_spec.md [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+**blend_id**<br>
+Reference ID for the blending mask used in this file. IDs should start at `0`.
+
+**filename**<br>
+Path to the image resource for the blending mask on the filesystem.
+
+There are two ways to specify a path: relative and absolute. Relative
+paths are relative to the location of the terrain definition file. They
+can only refer to image resources that are in the same modpack.
```
Comment (TheJJ) [inline doc/media/openage/terrain_format_spec.md:148]: i think for git-friendliness the best way to format paragraphs is one sentence per line, and no sentence wrapping at all. the rendered .md doesn't care anyway.

File: doc/media/openage/terrain_format_spec.md [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+scalefactor <factor>
+
+# layer and animation definitions
+# layers defined first will be overdrawn by later definitions
+layer <id> mode=off  position=<int>
+layer <id> mode=loop position=<int> time_per_frame=<float> replay_delay=<float>
+
+# definition of a terrain frames
+# these are iterated for an animation
+frame <layer_id> <image_id> <xpos> <ypos> <xsize> <ysize> blend_id=<int>
```
Comment (TheJJ) [inline doc/media/openage/terrain_format_spec.md:39]: why `blend_id` with keyword arg? also, which of the tiles in a blendmask to take?

**Review by TheJJ — CHANGES_REQUESTED** [review]

File: doc/media/openage/terrain_format_spec.md (hunk 1 above)
Comment (heinezen) [reply doc/media/openage/terrain_format_spec.md:132]: It was supposed to be just the PNG with the blend tiles, but you are right. We need to define the blending mask settings somehow. I'll come up with a solution...

File: doc/media/openage/terrain_format_spec.md (hunk 2 above)
Comment (heinezen) [reply doc/media/openage/terrain_format_spec.md:148]: I just had a lot of problematic experiences with text editors, so preserving readability for the pure Markdown was the goal here.

Does git actually care that much btw? The full blob is stored anyway when something changes and the few linebreaks should not make that much of a difference either.

File: doc/media/openage/terrain_format_spec.md (hunk 3 above)
Comment (heinezen) [reply doc/media/openage/terrain_format_spec.md:39]: `frame` is the full terrain texture in this case. Blending can be optional, that's why `blend_id` is an optional keyworded parameter. The tiles to take would depend on the hardcoded blendomatic algorithm.

File: doc/media/openage/terrain_format_spec.md (hunk 2 above)
Comment (TheJJ) [reply doc/media/openage/terrain_format_spec.md:148]: true, but the diff looks nicer :)
yea, was just a thought, we can leave it, it won't change that much anyway.

File: doc/media/openage/terrain_format_spec.md [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+...
+
+# sand.terrain
+...
+blendmask 0 "blend1.bltable" 10 3
+...
+
+-> grass.terrain's blending table has a higher blending priority.
+   Its blending table definition is therefore used when grass.terrain
+   and sand.terrain are adjacent to each other.
```
Comment (TheJJ) [inline doc/media/openage/terrain_format_spec.md:187]: for easier understandability i'd use the same mode here, and blend e.g. grass with forest, then only the priority should differ. the priority is also used (afair) to select which terrain tile flows on the other, i.e. which has flowing priority.

**Review by TheJJ — CHANGES_REQUESTED** [review]
getting there!

File: doc/media/openage/blendtable_format_spec.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+defined by the ID at the index.
+
+
+#### Example
+
+```
+blendtable
+0  0  0
+1  2  3
+23 42 1337
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/blendtable_format_spec.md:102]: Probably better to enclose matrices with `[ ]`. Maybe we should also use commas

File: doc/media/openage/blendtable_format_spec.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+
+## Quick Reference
+
+```
+# This is a blendtable configuration file
+# comments start with # and are ignored
+
+# file version
+version <version_no>
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/blendtable_format_spec.md]: fixed

File: doc/media/openage/blendmask_format_spec.md [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+
+
+## Quick Reference
+
+```
+# This is a blendtable configuration file
+# comments start with # and are ignored
+
+# file version
+version <version_no>
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/blendmask_format_spec.md]: fixed

File: doc/media/openage/blendmask_format_spec.md [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+imagefile 1 "/{aoe2_base}/sand.png"          # absolute path with modpack alias
+```
+
+Absolute paths are the only way to reference image resources from other
+modpacks.
+
+
+### `mask`
+
+Tells the renderer which mask it has to use.
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/blendmask_format_spec.md]: fixed

File: doc/media/openage/blendtable_format_spec.md [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+Lookup table for blendmasks blending two adjacent terrains.
+
+There has to be exactly one `blendtable` defined.
+
+Parameter | Type    | Optional | Default value
+----------|---------|----------|--------------
+matrix    | int[][] | No       | -
+
+**matrix**<br>
+A nxn matrix containing reference IDs for blending masks. This
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/blendtable_format_spec.md]: fixed

File: doc/media/openage/blendtable_format_spec.md [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+must be a square matrix. Columns are separated by spaces, rows
+are separated by newlines. IDs in the table must be defined in the
+same file by using the `blendmask` attribute.
+
+When two terrains are adjacent, the renderer selects the blending
+mask by looking it up in this table. For that purpose, it uses
+the `priority` ans `blend_mode` parameters of the `blendtable`
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/blendtable_format_spec.md]: fixed

File: doc/media/openage/blendtable_format_spec.md [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+attribute in the respective terrain format definitions.
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/blendtable_format_spec.md:88]: fixed

File: doc/media/openage/blendtable_format_spec.md [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+```
+blendmask 0 "blend0.blmask"        # blend0.blmask is in the same folder as the blendtable file
+blendmask 1 "./blend0.blmask"      # same as above, but more explicit
+blendmask 2 "media/blend1.blmask"  # blend1.blmask is in the subfolder media/
+```
+
+Absolute paths start from the (virtual) modpack root (the path where all
+modpacks are installed). They always begin with `/` followed by either
+a modpack identifier or a shortened modpack alias enclosed by `{}`. For
+information on modpack identifiers and aliases see the [TODO]() docs.
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/blendtable_format_spec.md]: fixed

File: doc/media/openage/terrain_format_spec.md [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+
+
+## Quick Reference
+
+```
+# This is a terrain configuration file
+# comments start with # and are ignored
+
+# file version
+version <version_no>
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/terrain_format_spec.md]: fixed

File: doc/media/openage/terrain_format_spec.md [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+```
+blendmask 0 "blend0.bltable" 10 3        # blend0.bltable is in the same folder as the terrain file
+blendmask 1 "./blend0.bltable" 10 3      # same as above, but more explicit
+blendmask 2 "media/blend1.bltable" 10 3  # blend1.bltable is in the subfolder media/
+```
+
+Absolute paths start from the (virtual) modpack root (the path where all
+modpacks are installed). They always begin with `/` followed by either
+a modpack identifier or a shortened modpack alias enclosed by `{}`. For
+information on modpack identifiers and aliases see the [TODO]() docs.
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/terrain_format_spec.md]: fixed

File: doc/media/openage/terrain_format_spec.md (hunk 4 above)
Comment (heinezen) [reply doc/media/openage/terrain_format_spec.md]: done

File: doc/media/openage/blendmask_format_spec.md [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+this factor can be used for textures that should retain high
+image quality with higher zoom levels.
+
+
+#### Example
+
+```
+# Assume the texture image has a size of 100x100
+scalefactor 1.0  # No scaling, 100x100 at default zoom
+scalefactor 2.0  # 50x50 at default zoom; 100x100 at 2x zoom
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/blendmask_format_spec.md]: rephrased the info a bit

File: doc/media/openage/blendmask_format_spec.md [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+       N-E-S-W
+
+          #
+        #####
+      ####0####
+    ####00000####
+      ####0####
+        #####
+          #
+```
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/blendmask_format_spec.md:317]: The ordering and IDs in `blendomatic.dat` are slightly different, so imo having the tiling info in this file again is better than just explaining the differences.

File: doc/media/openage/blendtable_format_spec.md [hunk 17]
```diff
@@ -0,0 +1,154 @@
+# Blendtable Format Specification
+
+**Format Version:** 1
+
+The openage blendtable format is a plaintext configuration file format for defining
+a blending mask table. The table can be used by terrain files to blend edges other
+terrains with an alpha mask. In the file, a blending lookup table and blending
+masks are declared.
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/blendtable_format_spec.md]: reworked

File: doc/media/openage/terrain_format_spec.md [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+
+**Format Version:** 1
+
+The openage terrain format is a plaintext configuration file format for defining
+terrain textures. It tells the openage renderer which image resources it has to load
+and how textures in these resources should be displayed.
+
+All attributes start with a defined keyword followed by parameter values. Some
+parameters have default values and are optional. The preferred file extension is
+`.terrain`.
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/terrain_format_spec.md:11]: Blending is optional, so I would leave this out in the file introduction.

File: doc/media/openage/terrain_format_spec.md [hunk 19]
```diff
@@ -0,0 +1,340 @@
...
+
+Absolute paths are the only way to reference image resources from other
+modpacks.
+
+**priority**<br>
+Decides which blending table of the two adjacent terrain textures is selected.
+The table referenced by the terrain with the highest priority value will be picked.
+If two adjacent terrains have equal priority, the blending table of the terrain
+with a lower x coordinate value is selected. If the x coordinate value is also
+equal, the blending table of the terrain with the lowest y coordinate is selected.
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/terrain_format_spec.md]: done

File: doc/media/openage/terrain_format_spec.md [hunk 20]
```diff
@@ -0,0 +1,340 @@
...
+imagefile 0 "grass.png"
+imagefile 1 "../../sand.png"
+imagefile 2 "/{aoe2_base}/graphics/grass.png"
+```
+
+
+### `blendtable`
+
+Defines a blending table that is used for looking up a blending mask. The
+blending mechanism is described in more detail in the [blendomatic](/doc/media/blendomatic.md)
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/terrain_format_spec.md:129]: done

File: doc/media/openage/blendtable_format_spec.md [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+1  2  3
+23 42 1337
+```
+
+
+### `blendmask`
+
+Defines a blending mask that is used to blend one terrain texture
+into a different terrain texture that is directly adjacent. The
+blending mechanism is described in more detail in the [blendomatic](/doc/media/blendomatic.md)
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/blendtable_format_spec.md]: added

File: doc/media/openage/blendmask_format_spec.md [hunk 22]
```diff
@@ -0,0 +1,318 @@
+# Blendmask Format Specification
+
+**Format Version:** 1
+
+The openage blendmask format is a plaintext configuration file format for defining
+a blending mask. The table can be used by terrain files to blend edges other
+terrains with an alpha mask.
```
[the remark this answers is not in the document]
Comment (heinezen) [reply doc/media/openage/blendmask_format_spec.md]: done

File: doc/media/openage/palette_format_spec.md [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+count      | int    | No       | -
+
+**count**<br>
+Number of colour values in the palette.
+
+
+#### Example
+
+```
+count 256
```
Comment (TheJJ) [inline doc/media/openage/palette_format_spec.md:79]: `entries`, not `count`

File: doc/media/openage/terrain_format_spec.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (TheJJ) [reply doc/media/openage/terrain_format_spec.md:11]: ah yea

**Review by TheJJ — CHANGES_REQUESTED** [review]
hopefully only one thing left

File: doc/media/openage/palette_format_spec.md (hunk 23 above)
Comment (heinezen) [reply doc/media/openage/palette_format_spec.md]: fixed

**Review by TheJJ — APPROVED** [review]
awesome!

**Pull request merged**